tinker_cookbook/renderers.py

```diff
@@ -553,8 +607,88 @@ def _end_message_token(self) -> int:
     def get_stop_sequences(self) -> list[int]:
         return [self._end_message_token]
 
+    def _parse_llama_tool_calls(self, content: str) -> list[ToolCall]:
+        """Parse tool calls from Llama 3 format.
+
+        Llama 3 uses: <function=function_name>{"arg": "value"}</function>
+        """
+        tool_calls: list[ToolCall] = []
+
+        for match in re.finditer(
+            r"<function=(\w+)>(.*?)</function>",
+            content,
+            re.DOTALL,
+        ):
+            func_name = match.group(1)
+            args_str = match.group(2).strip()
+            try:
+                # Validate JSON
+                json.loads(args_str)
+                tool_calls.append(
+                    ToolCall(
+                        function=ToolCall.FunctionBody(name=func_name, arguments=args_str),
+                    )
+                )
+            except json.JSONDecodeError:
+                continue
```

Comment:
Seems like this'll just silently fail and omit the tool call if it's not parsed correctly. Need to come up with a coherent policy for what to do when various kinds of tool errors occur.